lib/HLSL/DxilPreparePasses.cpp

```diff
@@ -399,50 +399,72 @@ class DxilFinalizeModule : public ModulePass {
+        OldEntryFunc->setName(Name);
+        Function *NewEntryFunc = StripFunctionParameter(OldEntryFunc, DM, FunctionDIs);
+        if (NewEntryFunc) {
+          DM.SetEntryFunction(NewEntryFunc);
+          OldEntryFunc->eraseFromParent();
         }
       }
     } else {
       std::vector<Function *> entries;
       // Handle when multiple hull shaders point to the same patch constant function
-      DenseMap<Function*,Function*> patchConstantUpdates;
+      MapVector<Function*, llvm::SmallVector<Function*, 2>> PatchConstantFuncUsers;
       for (iplist<Function>::iterator F : M.getFunctionList()) {
         if (DM.IsEntryThatUsesSignatures(F)) {
           auto *FT = F->getFunctionType();
           // Only do this when has parameters.
-          if (FT->getNumParams() > 0 || !FT->getReturnType()->isVoidTy())
+          if (FT->getNumParams() > 0 || !FT->getReturnType()->isVoidTy()) {
             entries.emplace_back(F);
+          }
+
+          DxilFunctionProps& props = DM.GetDxilFunctionProps(F);
+          if (props.IsHS() && props.ShaderProps.HS.patchConstantFunc) {
+            FunctionType* PatchConstantFuncTy = props.ShaderProps.HS.patchConstantFunc->getFunctionType();
+            if (PatchConstantFuncTy->getNumParams() > 0 || !PatchConstantFuncTy->getReturnType()->isVoidTy()) {
+              // Accumulate all hull shaders using a given patch constant function,
+              // so we can update it once and fix all hull shaders, without having an intermediary
+              // state where some hull shaders point to a destroyed patch constant function.
+              PatchConstantFuncUsers[props.ShaderProps.HS.patchConstantFunc].emplace_back(F);
+            }
+          }
         }
       }
-      for (Function *entry : entries) {
-        DxilFunctionProps &props = DM.GetDxilFunctionProps(entry);
-        if (props.IsHS()) {
-          // Strip patch constant function first.
-          Function* patchConstFunc = props.ShaderProps.HS.patchConstantFunc;
-          auto it = patchConstantUpdates.find(patchConstFunc);
-          if (it == patchConstantUpdates.end()) {
-            patchConstFunc = patchConstantUpdates[patchConstFunc] =
-                StripFunctionParameter(patchConstFunc, DM, FunctionDIs);
-          } else {
-            patchConstFunc = it->second;
-          }
-          if (patchConstFunc)
-            DM.SetPatchConstantFunctionForHS(entry, patchConstFunc);
+
+      // Strip patch constant functions first
+      for (auto &PatchConstantFuncEntry : PatchConstantFuncUsers) {
+        Function* OldPatchConstantFunc = PatchConstantFuncEntry.first;
+        Function* NewPatchConstantFunc = StripFunctionParameter(OldPatchConstantFunc, DM, FunctionDIs);
+        if (NewPatchConstantFunc) {
+          // Update all user hull shaders
+          for (Function *HullShaderFunc : PatchConstantFuncEntry.second)
+            DM.SetPatchConstantFunctionForHS(HullShaderFunc, NewPatchConstantFunc);
```

Comment:
@tex3d If you can have a look at this `SetPatchConstantFunctionForHS`. I think it has more issues because it's using a `set<Function*>` to keep track of the patch constant functions, but that could break down if multiple hull shaders have the same patch constant function. This is all for supporting `IsPatchConstantFunction()`, and I'm wondering if it'd be so bad to implement that by iterating over all the entry points and seeing if any is a hull shader pointing to that function. It would certainly be less error-prone and less bookkeeping.

Reply:
I see what you mean.  This code doesn't fix that problem though, right?  What about only removing from the set on RemoveFunction.  Is there a scenario where a function goes from being a patch constant function to something exported but no longer a patch constant function?  I don't think so...

Reply:
No, my code doesn't fix that, though it does fix the nondeterminism. I just spotted that additional issue while looking at that code,a nd I don't know if it causes actual bugs. To be correct, we would have to also iterate over all hull shader entry points and null out the `patchConstFunction` pointer in that `RemoveFunction`. And I can see more cases where we could get the `set<>` and the presence of non-null `patchConstFunction`s out of sync, which is why I'm thinking that maintaining that set may be more pain than it's worth.

Reply:
I filed issue #2115 on it. I'd like to get this in as-is so we at least have the nondeterminism fix because fixing the tracking of patch constant functions is a bit more involved.